File: core/play-guice/src/main/java/play/libs/akka/BehaviorProvider.java
/*
 * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package play.libs.akka;

import akka.actor.typed.Behavior;

import javax.inject.Provider;

// Must be an abstract class instead of an interface for TypedAkka#messageTypeOf to be able to
// extract the type of the parameterized type
public abstract class BehaviorProvider<Message> implements Provider<Behavior<Message>> {}

File: core/play-guice/src/main/scala/play/api/libs/concurrent/BehaviorProvider.scala
/*
 * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.libs.concurrent

import java.lang.invoke.MethodHandle

import akka.actor.typed.Behavior
import akka.annotation.ApiMayChange
import com.google.inject.{ Inject, Injector, Provider }
import javax.inject.Singleton

import scala.reflect.ClassTag

@Singleton
@ApiMayChange
final class BehaviorProvider[T: ClassTag](val methodHandle: MethodHandle) extends Provider[Behavior[T]] {
  @Inject private val guiceInjector: Injector  = null

  lazy val get = {
    val parameters = methodHandle
      .`type`()
      .parameterArray()
      .map { clazz =>
        guiceInjector.getInstance(clazz)
      }
    methodHandle.invokeWithArguments(parameters:_*).asInstanceOf[Behavior[T]]
  }
}
